File: haskell/interpreter/src/Lexer.hs
module Lexer (Lexer, new, tokenize, run) where

import Control.Applicative (Alternative ((<|>)))
import Control.Arrow (Arrow (first, second))
import Data.Char (isDigit, isLower, isSpace, isUpper)
import Data.Maybe (fromJust)
import Data.Tuple (swap)
import Token (TokenType (..))

newtype Lexer = Lexer String
    deriving (Show, Eq)

run :: String -> [TokenType]
run = fst . tokenize . new

new :: String -> Lexer
new = Lexer

tokenize :: Lexer -> ([TokenType], Lexer)
tokenize = inner []
  where
    inner (Illegal v : _) _ = error $ "Illegal character -- " ++ v
    inner acc@(Eof : _) l = (reverse acc, l)
    inner acc l = uncurry inner . swap . second (: acc) . nextToken $ l

advanceLexer :: Lexer -> Int -> Lexer
advanceLexer (Lexer input) v = Lexer $ drop v input

nextToken :: Lexer -> (Lexer, TokenType)
nextToken l@(Lexer input) = first (advanceLexer l) $ runHelper input
  where
    endOfFile v = case v of
        [] -> Just (Eof, 0)
        _ -> Nothing

    parseClean v =
        endOfFile v
            <|> parserSymbols v
            <|> keywords v
            <|> readIdentifier v
            <|> readInteger v
            <|> readString v
            <|> Just (Illegal v, 0)

    parser v len = second (+ len) <$> parseClean v

    runHelper = swap . fromJust . uncurry parser . skipWhitespace

isLetter :: Char -> Bool
isLetter c = or [fun c | fun <- [isUpper, isLower, (== '_')]]

keywords :: String -> Maybe (TokenType, Int)
keywords str = (,) <$> match s <*> Just (length s)
  where
    s = takeWhile isLetter str
    match v = case v of
        "fn" -> Just Function
        "let" -> Just Let
        "true" -> Just KTrue
        "false" -> Just KFalse
        "if" -> Just If
        "else" -> Just Else
        "return" -> Just Return
        _ -> Nothing

readBlock :: (a1 -> Bool) -> ([a1] -> a2) -> [a1] -> Maybe (a2, Int)
readBlock p f = helper . takeWhile p
  where
    helper [] = Nothing
    helper b = Just (f b, length b)

readIdentifier :: String -> Maybe (TokenType, Int)
readIdentifier = readBlock isLetter Ident

readInteger :: String -> Maybe (TokenType, Int)
readInteger = readBlock isDigit (Int . read)

readString :: String -> Maybe (TokenType, Int)
readString = outer
  where
    outer ('"' : c) = inner [] 1 c
    outer _ = Nothing

    inner acc count s = case s of
        [] -> Nothing
        ('\\' : '"' : cs) -> inner ('"' : acc) (count + 2) cs
        ('\\' : 'n' : cs) -> inner ('\n' : acc) (count + 2) cs
        ('\\' : 'r' : cs) -> inner ('\r' : acc) (count + 2) cs
        ('\\' : 't' : cs) -> inner ('\t' : acc) (count + 2) cs
        ('"' : _) -> Just (Str (reverse acc), count + 1)
        (curr : cs) -> inner (curr : acc) (count + 1) cs

parserSymbols :: String -> Maybe (TokenType, Int)
parserSymbols x = case x of
    ('=' : '=' : _) -> Just (Eq, 2)
    ('!' : '=' : _) -> Just (NotEq, 2)
    ('=' : _) -> Just (Assign, 1)
    ('+' : _) -> Just (Plus, 1)
    ('-' : _) -> Just (Minus, 1)
    ('*' : _) -> Just (Asterisk, 1)
    ('!' : _) -> Just (Bang, 1)
    ('/' : _) -> Just (Slash, 1)
    ('>' : _) -> Just (Gt, 1)
    ('<' : _) -> Just (Lt, 1)
    (';' : _) -> Just (Semicolon, 1)
    ('(' : _) -> Just (LParen, 1)
    (')' : _) -> Just (RParen, 1)
    (',' : _) -> Just (Comma, 1)
    ('{' : _) -> Just (LBrace, 1)
    ('}' : _) -> Just (RBrace, 1)
    ('[' : _) -> Just (LBracket, 1)
    (']' : _) -> Just (RBracket, 1)
    (':' : _) -> Just (Colon, 1)
    _ -> Nothing

skipWhitespace :: String -> (String, Int)
skipWhitespace = second length . swap . span isSpace

File: rust/interpreter/src/lexer.rs
use crate::token::Token;

pub fn tokenize<'a>(input: &'a str) -> Vec<Token<'a>> {
    Lexer::tokenize(input)
}

struct Lexer<'input> {
    input: &'input [u8],
}

type Return<'input> = Option<(Token<'input>, usize)>;

impl<'i> Lexer<'i> {
    fn tokenize(input: &'i str) -> Vec<Token<'i>> {
        let mut lexer = Self {
            input: input.as_bytes(),
        };
        let mut tokens = Vec::new();

        lexer.skip_whitespace();

        while let Some((token, by)) = lexer.next_token() {
            lexer.advance(by);
            tokens.push(token.clone());

            match token {
                Token::Eof | Token::Illegal(_) => break,
                _ => {}
            }
            lexer.skip_whitespace();
        }
        tokens
    }

    fn advance(&mut self, by: usize) {
        self.input = &self.input[by..];
    }

    fn next_token(&self) -> Return<'i> {
        let lst = [
            Lexer::eof,
            Lexer::parse_symbols,
            Lexer::parse_keywords,
            Lexer::parse_identifier,
            Lexer::parse_int,
        ];

        for l in lst {
            if let Some(a) = l(self) {
                return Some(a);
            }
        }

        Some((Token::Illegal(self.input[0]), 1))
    }

    fn eof(&self) -> Return<'i> {
        if self.input.len() <= 0 {
            Some((Token::Eof, 0))
        } else {
            None
        }
    }

    fn parse_identifier(&self) -> Return<'i> {
        self.take_while(Lexer::is_letter).map(|s| {
            let l = s.len();
            let ident = std::str::from_utf8(s).expect("Unable to unwrap ascii string");

            (Token::Ident(ident), l)
        })
    }

    fn parse_symbols(&self) -> Return<'i> {
        match self.input {
            &[b'=', b'=', ..] => Some((Token::Eq, 2)),
            &[b'!', b'=', ..] => Some((Token::Neq, 2)),
            &[b'!', ..] => Some((Token::Bang, 1)),
            &[b'=', ..] => Some((Token::Assign, 1)),
            &[b';', ..] => Some((Token::Semicolon, 1)),
            &[b'(', ..] => Some((Token::LParen, 1)),
            &[b')', ..] => Some((Token::RParen, 1)),
            &[b',', ..] => Some((Token::Comma, 1)),
            &[b'+', ..] => Some((Token::Plus, 1)),
            &[b'-', ..] => Some((Token::Minus, 1)),
            &[b'*', ..] => Some((Token::Asterisk, 1)),
            &[b'/', ..] => Some((Token::Slash, 1)),
            &[b'<', ..] => Some((Token::Lt, 1)),
            &[b'>', ..] => Some((Token::Gt, 1)),
            &[b'{', ..] => Some((Token::LBrace, 1)),
            &[b'}', ..] => Some((Token::RBrace, 1)),
            _ => None,
        }
    }

    fn parse_keywords(&self) -> Return<'i> {
        let keyword = self.take_while(Lexer::is_letter).map(|s| {
            std::str::from_utf8(s).expect("Unable to unwrap ascii string for the keywords")
        })?;

        let k = match keyword {
            "fn" => Token::Fun,
            "let" => Token::Let,
            "false" => Token::False,
            "true" => Token::True,
            "return" => Token::Return,
            "if" => Token::If,
            "else" => Token::Else,
            _ => return None,
        };

        Some((k, keyword.len()))
    }

    fn parse_int(&self) -> Return<'i> {
        let blk = self.take_while(Lexer::is_digit)?;
        let blk_in =
            std::str::from_utf8(blk).expect("Unable to unwrap ascii string for the keywords");

        Some((
            Token::Int(i64::from_str_radix(blk_in, 10).expect("Umable to convert correctly")),
            blk.len(),
        ))
    }

    fn skip_whitespace(&mut self) {
        let is_whitespace = |c: &&_| [b' ', b'\n', b'\t', b'\r'].contains(c);

        let it = self.input.iter().take_while(is_whitespace);
        let len = it.count();

        self.advance(len);
    }

    fn take_while<P>(&self, p: P) -> Option<&'i [u8]>
    where
        P: Fn(&u8) -> bool,
    {
        let it = self.input.iter().take_while(|&c| p(c));
        let len = it.count();

        if len == 0 {
            return None;
        }

        let (ident, _) = self.input.split_at(len);

        Some(ident)
    }

    fn is_digit(c: &u8) -> bool {
        c.is_ascii_digit()
    }

    fn is_letter(c: &u8) -> bool {
        c.is_ascii_alphabetic() || *c == b'_'
    }
}

#[cfg(test)]
mod test {
    use super::*;

    fn eq(input: &str, exp: &[Token]) {
        let got = tokenize(input);
        for (ref e, ref g) in std::iter::zip(exp, &got) {
            assert_eq!(e, g);
        }
    }

    #[test]
    fn test_next_token_simple() {
        let input = "=+(){},;";
        let exp = [
            Token::Assign,
            Token::Plus,
            Token::LParen,
            Token::RParen,
            Token::LBrace,
            Token::RBrace,
            Token::Comma,
            Token::Semicolon,
            Token::Eof,
        ];
        eq(input, &exp);
    }

    #[test]
    fn test_next_token_more_tokens() {
        let input = "let five = 5;
                     let ten = 10;
                     let add = fn(x, y) { x + y; }; 
                     let result = add(five, ten);";

        let exp = [
            Token::Let,
            Token::Ident("five"),
            Token::Assign,
            Token::Int(5),
            Token::Semicolon,
            Token::Let,
            Token::Ident("ten"),
            Token::Assign,
            Token::Int(10),
            Token::Semicolon,
            Token::Let,
            Token::Ident("add"),
            Token::Assign,
            Token::Fun,
            Token::LParen,
            Token::Ident("x"),
            Token::Comma,
            Token::Ident("y"),
            Token::RParen,
            Token::LBrace,
            Token::Ident("x"),
            Token::Plus,
            Token::Ident("y"),
            Token::Semicolon,
            Token::RBrace,
            Token::Semicolon,
            Token::Let,
            Token::Ident("result"),
            Token::Assign,
            Token::Ident("add"),
            Token::LParen,
            Token::Ident("five"),
            Token::Comma,
            Token::Ident("ten"),
            Token::RParen,
            Token::Semicolon,
            Token::Eof,
        ];

        eq(input, &exp);
    }

    #[test]
    fn test_next_token_complet() {
        let input = "
            let five = 5;
            let ten = 10;
            let add = fn(x, y) { x + y; };
            let result = add(five, ten); 
            !-/*5;
            5 < 10 > 5;
        ";

        let exp = [
            Token::Let,
            Token::Ident("five"),
            Token::Assign,
            Token::Int(5),
            Token::Semicolon,
            Token::Let,
            Token::Ident("ten"),
            Token::Assign,
            Token::Int(10),
            Token::Semicolon,
            Token::Let,
            Token::Ident("add"),
            Token::Assign,
            Token::Fun,
            Token::LParen,
            Token::Ident("x"),
            Token::Comma,
            Token::Ident("y"),
            Token::RParen,
            Token::LBrace,
            Token::Ident("x"),
            Token::Plus,
            Token::Ident("y"),
            Token::Semicolon,
            Token::RBrace,
            Token::Semicolon,
            Token::Let,
            Token::Ident("result"),
            Token::Assign,
            Token::Ident("add"),
            Token::LParen,
            Token::Ident("five"),
            Token::Comma,
            Token::Ident("ten"),
            Token::RParen,
            Token::Semicolon,
            Token::Bang,
            Token::Minus,
            Token::Slash,
            Token::Asterisk,
            Token::Int(5),
            Token::Semicolon,
            Token::Int(5),
            Token::Lt,
            Token::Int(10),
            Token::Gt,
            Token::Int(5),
            Token::Semicolon,
            Token::Eof,
        ];

        eq(input, &exp);
    }
}
